File: 2024/March/3-3/Approach2.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        def recursiveRemove(prev: ListNode, node: ListNode) -> int:
            nonlocal head
            
            if not node:
                return 1
            
            depth = recursiveRemove(node, node.next)
            if depth == n:
                if not prev:
                    head = node.next
                else:
                    prev.next = node.next

            return depth + 1

        recursiveRemove(None, head)
        return head

File: 2024/March/3-3/Approach2.java
public class Approach2 {
    public static void main(String[] args) {

    }
}

class Solution {
    private ListNode head;
    private int n;
    
    public ListNode removeNthFromEnd(ListNode head, int n) {
        this.head = head;
        this.n = n;
        
        recursiveRemove(null, head);

        return this.head;
    }

    private int recursiveRemove(ListNode previous, ListNode node) {
        if(node == null) {
            return 1;
        }

        int depth = recursiveRemove(node, node.next);
        if(depth == n) {
            if(previous == null) {
                head = node.next;
            }
            else {
                previous.next = node.next;
            }
        }

        return depth + 1;
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}